core/src/main/scala/org/apache/spark/util/JsonProtocol.scala

```diff
@@ -691,10 +724,42 @@ private[spark] object JsonProtocol {
   def accumulableInfoFromJson(json: JValue): AccumulableInfo = {
     val id = (json \ "ID").extract[Long]
     val name = (json \ "Name").extract[String]
-    val update = Utils.jsonOption(json \ "Update").map(_.extract[String])
-    val value = (json \ "Value").extract[String]
+    val update = Utils.jsonOption(json \ "Update").map { v => accumValueFromJson(name, v) }
```

Comment:
Looks like this will preserve backwards-compatibility as long as no users defined accumulators whose names start with our internal prefix. This seems like an okay compromise to me.